### ISSUE TITLE
Support for derived metrics

### ISSUE THREAD

**Issue by betodealmeida** [DataJunction/datajunction#113]
Given a metric `a` and a metric `b`, we should be able to define a metric `c` as:

```
SELECT a/b
```

**Comment by CircArgs** [comment]
Couple of questions for this one:

- Does this apply to nodes so you might have a metric node based on other metric nodes?
- Does this apply to metrics with `non-shared parents`?
    - If so, how might this handle dimensions?

**Comment by betodealmeida** [comment]
You ask good questions! :)

> Does this apply to nodes so you might have a metric node based on other metric nodes?

Right, the most common case is creating a new metric based on 2 other metrics with a simplified syntax that prevents code duplication. For example, imagine we have these 2 metrics:

```sql
-- page_views
SELECT COUNT(*)
FROM fact_actions
WHERE action_type='view' AND target='page'
```

```sql
-- page_clicks
SELECT COUNT(*)
FROM fact_actions
WHERE action_type='click' AND target='page'
```

If we now want to compute CTR (click through rate) we have at least 2 options to express that:

```sql
-- ctr option 1
SELECT
    COUNT(CASE WHEN action_type='click' THEN 1) / COUNT(CASE WHEN action_type='view' THEN 1)
FROM fact_actions
WHERE target='page'
```

```sql
-- ctr option 2
SELECT page_clicks / page_views
```

> Does this apply to metrics with non-shared parents? If so, how might this handle dimensions?

We can start with only metrics that share a parent, like the example above. For non-shared parents we need to define some extra metadata that tells us how to join the nodes. @agorajek implemented this in the original DJ at Facebook, he probably has better insight than I do.

**Comment by agorajek** [comment]
Yes... derived metric is probably a thing we'll need at some point. But I wouldn't go too fast on it. First of all you can make a derive metric using a regular metric approach, say:
```yaml
description: Click through rate.
type: metric
query: |-
  SELECT c.clicks / p.views AS value
  FROM page_clicks c JOIN page_views p
columns:
  value:
    type: DOUBLE
```

So, if you look at it this way, you should be make any metric out of other metric and even define special join conditions on them, because each metric node, has implicitly all the available dimension columns on them. So now, that we have the full functionality cover... we can decide if we need a short-hand mechanism to create them, with some assumptions in mind, e.g. that the participating lower level metrics will be JOINED using all their common dimensions. The last thing may be a little to risky to default to, but maybe not...
